### PR TITLE
ZhangShasha tree edit distance algorithm

### PR DESCRIPTION
**Pull request by SChudakov**
As suggested in #479 and #769 the JGraphT library currently lacks support for graph similarity measures algorithms. This PR adds the [Zhang-Shasha tree edit distance algorithm](http://www.grantjenks.com/wiki/_media/ideas:simple_fast_algorithms_for_the_editing_distance_between_tree_and_related_problems.pdf).

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/similarity/ZhangShashaTreeEditDistance.java [hunk 1]
```diff
@@ -0,0 +1,693 @@
...
+        }
+
+        /**
+         * Runs post-order DFS on {@code tree} starting at {@code treeRoot}.
+         * Assigns consecutive integer index to every traversed vertex and
+         * computes keyroots for {@code tree}.
+         *
+         * @param treeRoot root vertex of {@code tree}
+         */
+        private void computeKeyrootsAndMapping(V treeRoot) {
```
Comment (SChudakov) [inline jgrapht-core/src/main/java/org/jgrapht/alg/similarity/ZhangShashaTreeEditDistance.java:425]: This procedure uses linearized recursion. I will leave the original version of the recursive procedure to make the reviewing process more convenient.

```java
private int computeKeyrootsAndMapping(V v, List<V> stack, boolean isKeyroot) {
    stack.add(v);
    
    V vParent = null;
    if (stack.size() > 1) {
        vParent = stack.get(stack.size() - 2);
    }
    
    boolean isKeyrootArg = false;
    int lValue = -1;
    for (V vChild : Graphs.successorListOf(tree, v)) {
        if (vParent == null || !vChild.equals(vParent)) {
            int lVChild = computeKeyrootsAndMapping(vChild, stack, isKeyrootArg);
            isKeyrootArg = true;
            if (lValue == -1) {
                lValue = lVChild;
            }
        }
    }
    
    if (lValue == -1) { // this is a leaf
        lValue = currentIndex;
    }
    
    if (isKeyroot) {
        keyroots.add(currentIndex);
    }
    indexToVertexMap.set(currentIndex, v);
    indexToLMap.set(currentIndex, lValue);
    ++currentIndex;
    stack.remove(stack.size() - 1);
    
    return lValue;
}
```

File: jgrapht-core/src/main/java/org/jgrapht/alg/similarity/ZhangShashaTreeEditDistance.java [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.ToDoubleBiFunction;
+import java.util.function.ToDoubleFunction;
+
+/**
+ * Dynamic programming algorithm foe computing edit distance between trees.
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/similarity/ZhangShashaTreeEditDistance.java:33]: foe->for

File: jgrapht-core/src/main/java/org/jgrapht/alg/similarity/ZhangShashaTreeEditDistance.java [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+ * </ul>
+ *
+ * The algorithm assigns a cost to each of those operations which also depends
+ * on the vertices. The problem is then to compute a sequence of edit operations
+ * which transforms $T_1$ into $T_2$ and has a minimum cost over all such sequences.
+ * Here the cost of a sequence of edit operations is defined as sum of costs of
+ * individual operations.
+ *
+ * <p>
+ * The algorithm is base on a dynamic programming principle and assigns a label
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/similarity/ZhangShashaTreeEditDistance.java:65]: base->based

File: jgrapht-core/src/main/java/org/jgrapht/alg/similarity/ZhangShashaTreeEditDistance.java [hunk 4]
```diff
@@ -0,0 +1,698 @@
...
+ * keyroot in the first tree and $v_2$ is a keyroot in the second tree.
+ *
+ * @param <V> graph vertex type
+ * @param <E> graph edge type
+ * @author Semen Chudakov
+ */
+public class ZhangShashaTreeEditDistance<V, E> {
+
+    /**
+     * First tree for which the fistance is computed by this algorithm.
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/similarity/ZhangShashaTreeEditDistance.java:84]: fistance->distance

**Review by d-michail — COMMENTED** [review]
Looks good. Some minor typos.

**Comment by SChudakov** [comment]
Made the requested changes.

**Pull request merged**